### PR TITLE
Replace _iszero by iszerodefined && iszero

### PR DESCRIPTION
**Pull request by blegat**
Take 2 of https://github.com/JuliaLang/julia/pull/36194 after it was reverted in https://github.com/JuliaLang/julia/pull/38764.

What was missing in https://github.com/JuliaLang/julia/pull/36194 is that in case `iszero` is not working for a given type, we should drop the structure and fall back to an `Array`.
This PR defines a `iszerodefined` function that custom types implements in case `iszero` is well-defined even if `zero` is not defined.

### CODE REVIEW

**Comment by simeonschaub** [comment]
This might be ok, but we should probably do a PkgEval run, because this will now always produce dense matrices for non-numeric types that compare equal to `0`. I couldn't find any examples in Base that would be affected by this, but I could imagine that being relied upon in the wild.

**Comment by blegat** [comment]
Indeed, I have updated the PR with a different approach using `hasmethod` so that it also works for non-`Number` that implement `zero` and `==`

**Comment by simeonschaub** [comment]
I am not so sure about the usage of `hasmethod` here. This will always be a dynamic call, so it will probably be quite expensive for smaller arrays and cause type instabilities.

**Comment by blegat** [comment]
I have updated the PR so that it does not use `hasmethod`. It should be non-breaking as the new behavior is only enabled if a new method is implemented for `iszerodefined`. If `iszerodefined` is defined for a type `T`, it allows the structured broadcasting for `T` to work even if the type does not implement comparison with `Int`.

File: stdlib/LinearAlgebra/test/structuredbroadcast.jl [hunk 1]
```diff
@@ -206,6 +206,33 @@ end
 
 end
 
+struct Zero end
```
Comment (odow) [inline stdlib/LinearAlgebra/test/structuredbroadcast.jl:209]: Should this be
```julia
struct Zero36193 end
```
I can imagine other tests wanting to define a `Zero` struct.

**Comment by simeonschaub** [comment]
I guess the question is whether we want to make `iszerodefined` a public API and if yes, if it would make more sense for it to live in Base instead of LinearAlgebra. Alternatively, we could also make the current `_zero` into a more official API that we encourage people to overload for custom types - would that address your needs as well? Marking this for triage, so this can be discussed.

**Comment by blegat** [comment]
I agree that it would make sense to move `iszerodefined` to Base. It's not clear to me how `_iszero` could be moved to a an official API that is different to `iszero`, would it be "Same as `iszero` but falls back to comparing to the zero of `Int` instead of the zero of the type" ? I don't see a case where a type would return a different value from these different `iszero` and `_iszero` functions. Also, it seems to me that `_iszero` is more specific to this use case while `iszerodefined` would be useful in more cases.
Anyway, whatever is the decision, having either of the function part of the Base API would resolve our issue.

**Comment by JeffBezanson** [comment]
Lots of triage discussion on this. We discussed having a `Zero` type in Base, which already exists in some packages. It is similar to `Missing` in many ways. Most of us hesitate to go down this path however.

Should this only be true for Number? It also works for UniformScaling; not sure what else.

I find it a bit strange that a type would support `iszero` but not `zero`. Are there some good examples?

Stefan pointed out that to match e.g. `hastypemax` this should be `hasiszero` though that is a bit ugly. `haszero` is not too bad.

File: stdlib/LinearAlgebra/src/structuredbroadcast.jl [hunk 2]
```diff
@@ -104,9 +104,29 @@ function isstructurepreserving(::typeof(Base.literal_pow), ::Ref{typeof(^)}, ::S
...
-_iszero(n::Number) = iszero(n)
-_iszero(x) = x == 0
-fzeropreserving(bc) = (v = fzero(bc); !ismissing(v) && _iszero(v))
+"""
+    iszerodefined(T::Type)
+
+Return a `Bool` indicating whether `iszero` is well-defined for objects of type
+`T`. By default, this function returns `false` unless `T <: Number`. Note that
+this function may return `true` even if `zero(::T)` is not defined as long as
+`iszero(::T)` has a method that does not requires `zero(::T)`.
```
Comment (simeonschaub) [inline stdlib/LinearAlgebra/src/structuredbroadcast.jl:113]: I think we do actually want to require that `zero(::Type{T})` is defined, not just `zero(::T)`, which is an important distinction for arrays, among others. We never want to preserve structure if a function produces a vector for example:
```julia
julia> Diagonal([[1, 2], [3, 4]])[1, 2]
ERROR: MethodError: no method matching zero(::Type{Array{Int64,1}})
Closest candidates are:
  zero(::Type{Pkg.Resolve.VersionWeight}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Resolve/versionweights.jl:15
  zero(::Type{Pkg.Resolve.FieldValue}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Pkg/src/Resolve/fieldvalues.jl:38
  zero(::Type{Dates.Date}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Dates/src/types.jl:405
  ...
Stacktrace:
 [1] diagzero(::Diagonal{Array{Int64,1},Array{Array{Int64,1},1}}, ::Int64, ::Int64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/diagonal.jl:93
 [2] getindex(::Diagonal{Array{Int64,1},Array{Array{Int64,1},1}}, ::Int64, ::Int64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/diagonal.jl:89
 [3] top-level scope at REPL[19]:1
```

**Comment by odow** [comment]
> I find it a bit strange that a type would support iszero but not zero. Are there some good examples?

JuMP/MathOptInterface functions are examples of types that support `iszero` (check constant term is zero, and there are no affine terms), but either support `zero` and we regret it (https://github.com/jump-dev/JuMP.jl/issues/1151), or don't support zero (https://github.com/jump-dev/MathOptInterface.jl/issues/636). The main issue is that a lot of people write code assuming `zero(T)` is immutable, which isn't the case for JuMP (`zeros(T, dims...)` is a good example).

**Comment by blegat** [comment]
> I find it a bit strange that a type would support `iszero` but not `zero`. Are there some good examples?

In this case it's rather types `T` that are not subtypes of `Number`, support `iszero` but not `==(::T, ::Int)`.

**Comment by odow** [comment]
Bump. This is currently blocking using JuMP on Julia 1.6.

**Comment by simeonschaub** [comment]
I do think we eventually need a better solution here, but I am not yet convinced that as proposed, this is really better than what we currently have. In any case, as this is not a bugfix, I don't think we want to backport this to 1.6, so JuMP will probably still have to use the previous workaround in the mean time

**Comment by odow** [comment]
Okay, thanks. For reference, here is the work-around https://github.com/jump-dev/MutableArithmetics.jl/pull/70.

**Review by vtjnash — APPROVED** [review]
LGTM

**Pull request merged**